### PR TITLE
FAD-5655 Template and API Keys filters

### PR DESCRIPTION
**Pull request by jonambas**
- Adds FilterDropdown component. Added a usage guide as well.
- Implements filtering into Templates and API Keys list pages
- Adds EmptyState to Templates list page
- Adds custom header component for Template/Api Keys tables to set widths.

Lists are filtered through a selector BEFORE they hit the table collection component. Each list page will probably need its own filtering selector and filtering (redux-form wrapped) component.

TODO
- [x] Implement into API Keys list page
- [x] Tests
- [ ] Create ticket - handle `hasSubaccounts`

### CODE REVIEW

File: src/config/default.js [hunk 1]
```diff
@@ -27,6 +27,9 @@ const config = {
     supportEmail: 'support@sparkpost.com',
     billingEmail: 'billing@sparkpost.com'
   },
+  filters: {
+    minCount: 25
```
Comment (jonambas) [inline src/config/default.js]: New config key. Filters are show if list length is greater than this number.

File: src/components/reduxFormWrappers.js [hunk 2]
```diff
@@ -47,7 +47,7 @@ export const RadioGroup = ({ input, options, title }) => (
   </Radio.Group>
 );
 
-export const CheckboxWrapper = ({ input, ...rest }) => (
+export const CheckboxWrapper = ({ input, meta, ...rest }) => (
```
Comment (jonambas) [inline src/components/reduxFormWrappers.js]: Small bug I found, just stripping out meta before passing through to the Checkbox

File: src/components/collection/Collection.js [hunk 3]
```diff
@@ -74,7 +74,7 @@ class Collection extends Component {
         <OuterWrapper>
           <HeaderComponent />
           <BodyWrapper>
-            {this.getVisibleRows().map((row, i) => <RowComponent key={row[rowKeyName] || i} {...row} />)}
+            {this.getVisibleRows().map((row, i) => <RowComponent key={row[rowKeyName] ? `${row[rowKeyName]}-${i}` : i} {...row} />)}
```
Comment (jonambas) [inline src/components/collection/Collection.js]: Under the appteam account, there are a couple templates with the same `id` (idk). Should replace with `key={'${row[rowKeyName]}-${i}'}`

File: src/components/subaccountTypeahead/SubaccountTypeahead.js [hunk 4]
```diff
@@ -34,8 +34,7 @@ export class SubaccountTypeahead extends Component {
       (item) => `${item.name} ID: ${item.id}`
     ).map((item, index) => ({
       ...getItemProps({ item, index }),
-      content: <Item name={item.name} id={item.id} />,
-      highlighted: highlightedIndex === index
```
Comment (jonambas) [inline src/components/subaccountTypeahead/SubaccountTypeahead.js]: Removed `highlighted` prop from MB, replaced with `selected` which adds an `<Icon name='Check'>`

File: src/pages/api-keys/components/TableComponents.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+const columns = [
+  { label: 'Name', width: '40%' },
+  { label: 'Key', width: '20%' },
+  { label: 'Subaccount', width: '20%' }
+];
+
+export const TableHeader = () => (
+  <thead>
+    <Table.Row>
+      {columns.map(({ label, ...rest }) => <Table.HeaderCell key={label} {...rest}>{label}</Table.HeaderCell>)}
```
Comment (jonambas) [inline src/pages/api-keys/components/TableComponents.js]: Should probably just make this the default TableHeader for the TableCollection

File: src/components/collection/Collection.js [hunk 6]
```diff
@@ -64,17 +64,17 @@ class Collection extends Component {
...
     } = this.props;
 
     return (
       <div>
         <OuterWrapper>
-          <HeaderComponent />
+          {headerComponent}
           <BodyWrapper>
-            {this.getVisibleRows().map((row, i) => <RowComponent key={row[rowKeyName] || i} {...row} />)}
+            {this.getVisibleRows().map((row, i) => <RowComponent key={`${row[rowKeyName]}-${i}`} {...row} />)}
```
Comment (jasonrhodes) [inline src/components/collection/Collection.js]: why this change? if the key is unique why does it need the `i` appended?

File: src/components/collection/Collection.js [hunk 7]
```diff
@@ -64,17 +64,17 @@ class Collection extends Component {
...
+      headerComponent,
       outerWrapper: OuterWrapper = PassThroughWrapper,
       bodyWrapper: BodyWrapper = PassThroughWrapper
     } = this.props;
 
     return (
       <div>
         <OuterWrapper>
-          <HeaderComponent />
+          {headerComponent}
```
Comment (jasonrhodes) [inline src/components/collection/Collection.js]: not sure I fully understand this change ... it _seems_ better to pass components around and not JSX strings?

File: src/components/filterDropdown/FilterDropdown.js [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+          trigger={<TextField prefix={prefix} value={displayValue} readOnly suffix={<Icon name='CaretDown'/>} />}>
+          <ActionList actions={actions} />
+        </Popover>
+        {this.renderCheckboxes()}
+      </div>
+    );
+  }
+}
+
+FilterDropdown.propTypes = {
```
Comment (jasonrhodes) [inline src/components/filterDropdown/FilterDropdown.js]: Are we using propTypes in this codebase? I don't know much about them so I don't actually have an opinion beyond "so long as they aren't included in the prod bundle at all" -- if we're using them should we cut a ticket to apply them everywhere?

File: src/components/reduxFormWrappers.js (hunk 2 above)
Comment (jasonrhodes) [reply src/components/reduxFormWrappers.js]: Can you use `_.omit` for this? I was doing this exact thing somewhere else but it got a bit sticky and once we fix our linting this is going to get nailed for unused variables.

Something like:
```js
({ input, ...rest }) => {
  const checkBoxProps = _.omit(rest, 'meta');
  // etc
```

**Comment by jasonrhodes** [comment]
This is good. I'd like to see if we can put the filters inside of the TableCollection as some sort of declarative prop somehow, so we don't have to pre-filter the list given to TableCollection. It could render the filters (if necessary) and then apply the filtering to its own set of rows, which all seems very natural.

File: src/components/collection/Collection.js (hunk 6 above)
Comment (jonambas) [reply src/components/collection/Collection.js]: On the appteam account, there are templates with the same ID - not sure where they came from: 
screen shot 2017-10-17 at 2 31 14 pm

File: src/components/collection/Collection.js (hunk 7 above)
Comment (jonambas) [reply src/components/collection/Collection.js]: I removed the custom header components from the list pages, so I can revert. But, IMO it looks a little cleaner.

```js
<TableCollection headerComponent={<Component {...props}/>}

// or

<TableCollection headerComponent={() => <Component {...props}/>}

// or

<TableCollection headerComponent={Component} /> // but how to add props?
```

File: src/components/filterDropdown/FilterDropdown.js (hunk 8 above)
Comment (jonambas) [reply src/components/filterDropdown/FilterDropdown.js]: We don't have to apply them everywhere. I think for these complicated components its just useful to say "hey you're not using this correctly"

**Comment by jonambas** [comment]
I wasn't really sure how to get it integrated into the TableCollection. To get filters working you need:
1. a redux-form filters component
2. a selector to filter the list (with a selector to get form values)

Would still have to pass those to the collection component, which would only really be doing the `list.length > config.filters.minCount` check

File: src/components/subaccountTypeahead/SubaccountTypeahead.js (hunk 4 above)
Comment (jonambas) [reply src/components/subaccountTypeahead/SubaccountTypeahead.js]: Actually - should add back in. Key events don't show up (oops)

File: src/components/reduxFormWrappers.js (hunk 2 above)
Comment (jonambas) [reply src/components/reduxFormWrappers.js]: Realized this component doesn't pass through `error`, will use meta like the other wrapper components:
```js
export const CheckboxWrapper = ({ input, meta, ...rest }) => {
  const { active, error, touched } = meta;
  return (
    <Checkbox
      id={input.name}
      {...input}
      error={!active && touched && error ? error : undefined}
      {...rest}
    />
  );
};
```

File: src/components/collection/Collection.js [hunk 9]
```diff
@@ -1,18 +1,23 @@
...
-import Pagination from './Pagination';
+import Pagination, { defaultPerPageButtons } from './Pagination';
+import CollectionFilter from './Filter';
+import { objectSortMatch } from 'src/helpers/sortMatch';
 
 const PassThroughWrapper = (props) => props.children;
+const NullComponent = () => null;
 
-class Collection extends Component {
+export class _Collection extends Component {
```
Comment (jonambas) [inline src/components/collection/Collection.js]: Why underscore?

File: src/components/collection/Pagination.js [hunk 10]
```diff
@@ -23,7 +25,7 @@ export default class CollectionPagination extends Component {
   }
 
   renderPerPageButtons() {
-    const { data, perPage, perPageButtons = [10, 25, 50, 100], onPerPageChange } = this.props;
+    const { data, perPage, perPageButtons = defaultPerPageButtons, onPerPageChange } = this.props;
```
Comment (jonambas) [inline src/components/collection/Pagination.js]: Nitpick but we should try to use `defaultProps` instead of defaulting in a spread

File: src/components/filterDropdown/tests/__snapshots__/FilterDropdown.test.js.snap [hunk 11]
```diff
@@ -0,0 +1,139 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`renders correctly 1`] = `
+<div>
+  <t
```
Comment (jonambas) [inline src/components/filterDropdown/tests/__snapshots__/FilterDropdown.test.js.snap]: Will probably need to update MB and these tests

File: src/config/default.js (hunk 1 above)
Comment (jonambas) [reply src/config/default.js]: Can remove i think?

File: src/pages/api-keys/ListPage.js [hunk 12]
```diff
@@ -89,13 +78,14 @@ export class ListPage extends Component {
   }
 
   render() {
-    const { error, loading, newKey } = this.props;
+    const { error, loading, newKey, count } = this.props;
 
     return (
       <Layout.App loading={loading}>
```
Comment (jonambas) [inline src/pages/api-keys/ListPage.js]: Will need to update with layout & loading changes

File: src/pages/api-keys/components/Filters.js [hunk 13]
```diff
@@ -0,0 +1,41 @@
+import React from 'react';
```
Comment (jonambas) [inline src/pages/api-keys/components/Filters.js]: Can delete file

File: src/pages/templates/tableConfig.js [hunk 14]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+import { Link } from 'react-router-dom';
+import { format } from 'date-fns';
+
+export const getRowData = ({ published, id, name, last_update_time, subaccount_id, shared_with_subaccounts }) => [
+  <Link to={`/templates/edit/${id}`}>{name}</Link>,
+  id,
+  published ? 'Published' : 'Draft',
+  format(last_update_time, 'MMM D, YYYY [at] h:mma'),
```
Comment (jonambas) [inline src/pages/templates/tableConfig.js]: if theres a test that renders these rows, might need to mock `format` so these pass in different timezones. Should be okay with `shallow`

File: src/selectors/templates.js [hunk 15]
```diff
@@ -1,7 +1,60 @@
...
+import sortMatch from 'src/helpers/sortMatch';
 
-export {
-  templateById
+export const getTemplates = (state) => state.templates.list;
+export const getTemplateById = (state, props) => state.templates.byId[props.match.params.id] || { draft: {}, published: {}};
+
+const getFilters = (state) => state.form.templateFilters && state.form.templateFilters.values;
+
+export const getfilterTemplates = createSelector(
```
Comment (jonambas) [inline src/selectors/templates.js]: Can remove as well if we aren't using filter dropdowns?

File: src/components/collection/Collection.js (hunk 9 above)
Comment (jasonrhodes) [reply src/components/collection/Collection.js]: I know we aren't doing this everywhere but it always makes me a little nervous when you can do both of these:

```js
import Collection from 'src/components';
import { Collection } from 'src/components';
```

So the `_prefix` seemed like maybe a good way to signify an export that's sort of meant to be internal-only for testing etc... 

Could be convinced to not care.
